### ISSUE TITLE
jest ts with jest-expo fails with timeout

### ISSUE THREAD

**Issue by PavelPolyakov** [expo/expo#4293]
### Steps to Reproduce
My expo project is a typescript one. Therefore I also want to have tests in `ts`. For tests I use `jest`.
Everything is configured more or less according to this guide:

What happens when I process `ts` tests with `jest-expo` - they stuck whenever any of the assertion fails.

Here is the reproduction:
![image](https://user-images.githubusercontent.com/839290/58257698-aab0ca00-7d71-11e9-8e55-7e7400e53124.png)

On the left side I'm running simple `ts` test, on the right side I run `ts` tests with `jest-expo`.

### Expected Behavior
If assertion fails, you just see it in the cli. Like it happens on the left side, with simple `ts`.

### Actual Behavior
If any of assertion fails - everything stops and `jest` fails because of the timeout.

### Reproducible Demo
Here is the full reproduction repo:
https://github.com/PP-etc/jest-timeout

Any hints how to put it on the correct rails?

FYI @sjchmiela

Regards,

**Comment by ide** [comment]
I'm not sure what the issue is and I'd recommend debugging the internals of Jest and stepping through the stack frames to understand what is going on. If you identify the root cause, we can look into fixing it. For what it is worth, we use ts-jest in this repo.

**Comment by SimenB** [comment]
`expect` throws, which means you never call `done`. You should keep to promises :)

**Comment by ide** [comment]
Great point. Perhaps Jest should blow up if you do `async function (done)`.

**Comment by SimenB** [comment]
Yeah, probably. You can do `await new Promise` if you need callback. Personally I wanna remove done completely but that'll probably break fb in a bajillion ways :p

We have https://github.com/jest-community/eslint-plugin-jest/blob/master/docs/rules/no-test-callback.md

**Comment by ide** [comment]
@SimenB Jest could keep `done` and I just meant that Jest could check if `isPromise(specFn()) && specFn.length > 0`.

**Comment by ide** [comment]
@PavelPolyakov I'm not sure why Jest hangs only with jest-expo (feel free to investigate) but I would see if removing `done` helps.

**Comment by PavelPolyakov** [comment]
Well, yes, obviously it stucks because `done` is not called and that promise is not resolved. But why only with `jest-expo` - that's the question.
I can't remove `done`, because my real tests are `async` and I need `it` function to be `async` as well.

one can also wrap each test in `try/catch` and make sure `done` is called, this also helps, but this is not how it is suppose to work. Therefore I still consider this as a bug, the fact that it only appears if `jest-expo` is in the game makes me think that it's `jest-expo` bug.

I'll try to investigate more, but I don't even know where to start :)

**Comment by PavelPolyakov** [comment]
I've also made another experiment and tested the behaviour with another preset:
https://github.com/PP-etc/jest-timeout/tree/master/ts-with-another-preset

This one works fine, fails because of an error.
So the current situation is only present with `jest-expo` OR `jest-expo` + `react-native` transformation.

**Comment by ide** [comment]
We agree this is a bug and not intended behavior but don't know the root cause, and identifying that would be helpful.

**Comment by webinoo** [comment]
This seems like it was resolved in some versions of jest-expo between 32.0.1 and 44.0.1.

My issue was that the Promises were not being resolved and when I was testing redux async thunk, the test got stuck on first await. I separated the issue and used @PavelPolyakov repo and added my test

```
  test("async sample", async () => {
    const asyncMock = jest.fn().mockResolvedValue(42);
    await expect(asyncMock()).resolves.toEqual(42);
  });
```

The same was happening that it got stuck in `ts-with-jest-expo` folder but working in `ts` folder.
Currently, it works again, if you take the repo
- upgrade jest-expo to 44.0.1
- remove property `preset: "jest-expo"` from jest.config.js

Thank you @PavelPolyakov for help on separating and minimizing the issue.
Correct me if I'm wrong, but I think this Issue can be closed now.
